### ISSUE TITLE
[red-knot] Duplicate diagnostics in unpacking assignment

### ISSUE THREAD

**Issue by dhruvmanila** [astral-sh/ruff#16514]
### Summary

Consider the following code:
```py
class Foo():
    pass

a, b = Foo().nonexistant
```

Here, the RHS of the assignment has an error `unresolved-attribute` which is being shown 3 times (2 + 1 where 2 is the number of variables involved in the unpacking):

<img width="1043" alt="Image" src="https://github.com/user-attachments/assets/e6be58cc-cc5e-4079-814c-de25191bc10d" />

This is because the standalone expression is being inferred _outside_ the unpacking in `infer_assignment_definition` (and similarly in `infer_for_statement_definition`). This inference should only be done for non-unpack targets.

Fixing this would lead to another bug which is that the diagnostics raised during the unpacking query isn't being merged into the outer type inference builder.
